### DISCUSSION TITLE
如何使用计划任务打通简悦导出到 Logseq 导入的最后一步

### DISCUSSION THREAD

**Discussion by Guan810** [Kenshin/simpread#3751] [服务]
# 自动化方案 

当导出时，复制文件到 Logseq 的文件夹中，然后追加标题信息到当天日记中去。若当天日记不存在，就新建一个再追加。

# 事情起源

事情起源于配置简悦与 Logseq 联动自动化的过程中。由简悦定制导出的 Markdown ，为了能够添加到 Logseq 的日记中，往往时使用反链接的方式，对于有精神洁癖的人而言，未免有些难受。

于是开始寻找更好的方式，比如导出的 Markdown 可以自动将文件放入 Logseq ，且把标题记录在当天日记中去。

查看 Logseq 的日记文件可以发现， Logseq 仅保存当天有内容的日记，且名称就是当天的日期。而没有内容的日记是没有对应的文件。

![image_1650078724482_0.png](https://s2.loli.net/2022/04/18/yOHm3TIjYUZB7CR.png)

因此我们的自动化操作就很简单了：当导出时，复制文件到 Logseq 的文件夹中，然后追加标题信息到当天日记中去。若当天日记不存在，就新建一个再追加。

# 技术选型

作为精神洁癖，笔者奉行的标准就是能用现成的绝不下载软件，功能越单一的软件哪怕界面简陋只要我看不见那也很不错。

那么还有什么比 Windows 自带的计划任务更简洁的了呢？

计划任务的本质是触发后执行程序，这里完全可以使用各种脚本语言，如python。

有了这两个已经足够设置一个定时任务了，但是如果想要实现“每次修改再触发程序”这个功能，还需要配置一下事件日志。

# 事件日志配置

桌面级的 Windows 中默认是不开启审核的，因此我们需要手动配置一下。

首先找到简悦导出的文件夹，在右键菜单中选择属性，然后依次 `安全 > 高级 > 审核 > 添加`，在选择主体中`高级 > 立即查找`，选择自己就可以了。在基本权限中选择自己需要的就可以，我的选择是“创建文件/写入数据、创建文件夹/附加数据、删除、删除子文件夹及文件”这几项。

理论上全选了也没啥，可能就是日志文件多了几个KB而已。

之后需要在组策略中开启审核。在运行中输入 `secpol.msc` 打开组策略，选择 `本地策略 > 审核策略 > 审核对象访问`，选择 `审核成功` 即可。

![image_1650080319635_0.png](https://s2.loli.net/2022/04/18/5OuHLSj3d7IwoyG.png)

至于家庭版如何开启组策略，相关教程可以在 [这里](http://www.xitongzhijia.net/xtjc/20220111/237843.html) 找到。

# 计划任务配置

- 计划任务配置可以在 `开始 > 计算机管理 > 计划任务程序` 找到，使用创建基本任务就可以按照流程配置了。

- 其实还可以从事件查看器中直接创建。在导出的文件夹中新建个 txt文档（这是为了触发日志），在 `事件查看器 > Windows 日志 > 安全` 中选择查找，输入你的导出文件夹名，就可以找到对应事件。在右侧下方应该改有一个选项“将任务附加到此事件”，一路确认下去到最后一步打开属性进行下一步设置。

- 计划任务属性中的触发器可以修改触发条件，操作中可以修改触发的程序或批处理文件。选择 `触发器 > 编辑 > 自定义 > xml > 手动编辑`，之后在文本框中输入以下文本，需要修改同步助手的路径，我会在之后介绍。

```
<QueryList>
    <Query Id="0" Path="Security">
    <Select Path="Security">
    *[System[Provider[@Name='Microsoft-Windows-Security-Auditing'] and (Level=4 or Level=0) and Task = 12800 and (band(Keywords,9007199254740992)) and (EventID=4663)]] 
    and 
    *[EventData[Data[@Name='ProcessName'] and (Data='Path/To/EXE')]]</Select>
    </Query>
</QueryList>
```

![image_1650081092731_0.png](https://s2.loli.net/2022/04/18/p12vHOWm4k6FUDc.png)

- 同步助手的路径并非正常的安装目录，而是需要在日志中寻找。在按照第二步找到对应事件后，查看详情可以找到“进程名”，那个就是需要填充在代码中的值，直接复制即可。

- 在完成以上步骤后，还需要在常规中修改运行任务的用户账户，修改方式和配置事件日志的选择主题是一样的。

- 这里可能是 Windows 的 Bug，不修改就会报错，修改了下回还得修改，反正就是每次动都得改这个。

- 之后可以选择操作中的运行程序为 cmd，方便我们测试一下。保存后试一下导出到本地，看看有没有窗口弹出。

# Python 文件

- Python 本身教程很多，也很强大，这里我就不再赘述了。只要明确任务，就能够实现很多花活。

- 为了解决随着本地文件越多，复制越多的情况，最好是根据文件的最后修改日期进行一个判断。

- 最后附一个我自己的 Python 文件做参考：

```
import codecs
import datetime
import os
import shutil
import sys

def main_process(origin_path, target_path, journals_path, time):
    copy_date = datetime.datetime.now() - datetime.timedelta(minutes=time)
    add_date = datetime.date.today()
    journal = os.path.join(journals_path, "%d_%d_%d.md" % (add_date.year, add_date.month, add_date.day))
    print(copy_date, add_date, journal)

    for f in os.listdir(origin_path):
        origin_file = os.path.join(origin_path, f)
        target_file = os.path.join(target_path, f)
        print(origin_file, target_file)
        origin_mtime = datetime.datetime.fromtimestamp(os.path.getmtime(origin_file))
        if origin_mtime > copy_date:
            mark = 0
            if os.path.exists(target_file):
                mark = 1
                target_mtime = datetime.datetime.fromtimestamp(os.path.getmtime(target_file))
                if target_mtime.date() == add_date:
                    mark = 2

            print(mark)
            shutil.copyfile(origin_file, target_file)
            if mark != 2:
                new_journal = 0
                if os.path.exists(journal):
                    new_journal = 1
                with codecs.open(journal, "a", encoding="utf-8") as j:
                    if new_journal == 1:
                        j.write("\n")
                    j.write("- [[%s]]" % (os.path.splitext(f)[0]))


if __name__ == '__main__':
    main_process(sys.argv[1], sys.argv[2], sys.argv[3], int(sys.argv[4]))
```

- 四个参数分别为简悦导出文件夹、复制到 Logseq 的文件夹、 Logseq 日记文件夹和最后修改日期与当前时间相差的分钟数。最后一个参数为15时意味着只有最早15分钟前导出过的文件才会被复制。

- 在计划程序中，调用程序为 python，参数为 py 文件地址和以上四个参数

# pythonw 静默执行

参考文章 [Windows 计划任务执行如何静默执行python程序](https://blog.csdn.net/qq_21148843/article/details/104255854)

以上配置结束后，每次执行都会闪一个 Dos 界面，为了能够静默执行可以将程序改为 `pythonw.exe`（与 `python.exe` 同文件夹 ），对应的将 py 文件后缀改为 pyw。

## 补充

> 以下内容为简悦官方补充说明。

如果将网页的标注导入到 Logseq 的话，可以看这篇文章 [简悦 + Logseq 个人知识库搭建 | 从零开始完全指南（适合无任何经验的小白用户）](https://zhuanlan.zhihu.com/p/489510457)

简悦与 Logseq 的更多联动方案可以 [看这里](https://github.com/Kenshin/simpread/discussions?discussions_q=label%3Alogseq)。

**Comment by Kenshin** [comment]
哈哈，不错的方式，谢谢分享。 👏 

我知道你的文章是通过 Logseq 导出的，所以顺道修改了下文章结构。

**Comment by CenBoMin** [comment]
我有个想法不知道能不能实现，坚果云可以锁定logseq的日记资料夹同步，然后只要增加今天日期md文件到坚果云，让坚果云同步到本地。

我看了下简悦开放的API可以查询日期和标题和高亮的标注，如果可以用nodejs结合资料库的方式，定时任务请求API然后筛选出来在重组生成一个md文档，最后在上传坚果云里面也许可行。

坚果云这块不太熟悉，不知道怎么自动上传文档上去和能不能锁定特定的同步资料夹上传，不过我想原理应该和简悦用坚果云webdev同步配置文档是一样的

**Comment by Kenshin** [comment]
嗨，各位。

简悦官方针对新用户或不喜欢配置的用户，发布了一个新功能：**Logseq 配置库**。

细节请看下面，几分钟就能配置成完成。

***

![img](https://cdn.jsdelivr.net/gh/23784148/upload-images@main/typora/20221007_1665118121.png)

> 针对新用户的一步到位的 Logseq 配置库。

## 🗂 配置库

[简悦 · 配置库](https://www.yuque.com/kenshin/simpread/ds8zk0) 是简悦官方推出的一套针对新用户的极简配置方案，方便新用户用最快的方式使用简悦的各种高级服务，配置库内置了常用的双链笔记用法，如：Notion、Obsidian、Logseq、Roam Research，同时包含了简悦在阅读模式上的一些常规插件：Live Editor、题图、Safari 阅读模式等。

## 💡 特点

用户仅需要下载对应的配置库，修改里面的 UID 为自己的高级账户 UID，即可实现简悦的各种高级操作。

## 💁‍♂️ 受众

新用户或者不想使用手动配置的用户。

## 📝 如何使用

1️⃣ 下载

2️⃣ 修改为自己的 UID

3️⃣ 将必要的文件夹复制粘贴到相应位置

✅ Done！

## 📦 Logseq 配置库

因为 Logseq 与 Obsidian 在插件 API 的支持度有限，所以并没有轻量级使用方案，根据 Logseq 支持的文件格式区别两种方案。

这两种方案均需要同步助手，几乎无任何配置，支持本地快照，自动同步标注，可在 Logseq 中标注，适合 Logseq + 简悦的重度用户。

### 🔌 功能

1️⃣ 支持在 Logseq 中标注

2️⃣ 加入稍后读，自动保存本地快照

3️⃣ 加入标注时，自动将标注内容导入本地快照和 Logseq

4️⃣ 改变稍后读或标注的元数据时，自动将改动的内容保存到本地快照和 Logseq

### 📖 区别

![img](https://imgs.zhubai.love/66078e55f4764b3eaf1a7c1f7cc64c28.png)

### 📙 Org Mode 版

🔗 [Github](https://github.com/Kenshin/simpread-configs/blob/main/logseq@orgmode/Getting%20Started.md) | [语雀](https://www.yuque.com/kenshin/simpread/mdgp9n)

### 📗 Markdown 版

🔗 [Github](https://github.com/Kenshin/simpread-configs/blob/main/logseq@md/Getting%20Started.md) | [语雀](https://www.yuque.com/kenshin/simpread/tukkeo)
